File: go/internal/pairing/code.go
// go/internal/pairing/code.go
package pairing

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/url"
)

// NewToken returns a fresh 16-byte (128-bit) single-use pairing token.
func NewToken() []byte {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	return b
}

type codePayload struct {
	S string `json:"s"` // signal URL
	T string `json:"t"` // token hex
}

// EncodeCode produces a self-contained, copy-pasteable pairing code.
func EncodeCode(signalURL string, token []byte) string {
	data, _ := json.Marshal(codePayload{S: signalURL, T: hex.EncodeToString(token)})
	return base64.RawURLEncoding.EncodeToString(data)
}

// DecodeCode parses a pairing code into its signal URL and token.
func DecodeCode(code string) (signalURL string, token []byte, err error) {
	data, err := base64.RawURLEncoding.DecodeString(code)
	if err != nil {
		return "", nil, fmt.Errorf("bad pairing code: %w", err)
	}
	var p codePayload
	if err := json.Unmarshal(data, &p); err != nil {
		return "", nil, fmt.Errorf("bad pairing code: %w", err)
	}
	token, err = hex.DecodeString(p.T)
	if err != nil || len(token) != 16 {
		return "", nil, fmt.Errorf("bad pairing code token")
	}
	// Validate the relay URL before it ever drives DialPair/websocket.Dial or is
	// persisted. The relay is untrusted by design (Noise authenticates the peer,
	// not the transport), but a scanned/pasted code is attacker-controlled, so
	// reject anything that is not a real http(s) origin: no javascript:/data:/
	// file: schemes, no plain-http to a public host (downgrade), no garbage that
	// would corrupt the ws:// derivation. Mirrors web/src/pairing/code.js
	// (validSignalURL) — fail closed identically so a CLI never silently dials a
	// hostile transport. http is allowed only for localhost dev.
	if !validSignalURL(p.S) {
		return "", nil, fmt.Errorf("bad pairing code signal URL")
	}
	return p.S, token, nil
}

// validSignalURL reports whether s is a relay URL we are willing to dial. It
// mirrors validSignalURL in web/src/pairing/code.js: require an http(s) URL;
// https is always allowed; plain http only for localhost / 127.0.0.1; everything
// else (javascript:/data:/file:, other schemes, malformed, schemeless) is
// rejected. url.Parse lowercases the scheme, matching the JS URL.protocol check.
func validSignalURL(s string) bool {
	u, err := url.Parse(s)
	if err != nil {
		return false
	}
	host := u.Hostname()
	if host == "" {
		// JS `new URL` throws on an authority-less URL like "https:" / "https://";
		// fail closed to match (and a hostless URL can't drive a real ws:// dial).
		return false
	}
	switch u.Scheme {
	case "https":
		return true
	case "http":
		return host == "localhost" || host == "127.0.0.1"
	default:
		return false
	}
}

// pskFromToken derives the 32-byte Noise PSK (domain-separated from roomID).
func pskFromToken(token []byte) []byte {
	h := sha256.Sum256(append([]byte("terminal-relay/pair/psk"), token...))
	return h[:]
}

// RoomID derives the public rendezvous id (domain-separated from the psk).
func RoomID(token []byte) string {
	h := sha256.Sum256(append([]byte("terminal-relay/pair/room"), token...))
	return hex.EncodeToString(h[:16])
}

File: web/src/pairing/code.js
// web/src/pairing/code.js
// Pairing-code / roomID / psk derivation for terminal-relay browser pairing.
// Must match the Go reference (go/internal/pairing/code.go) byte-for-byte.
import { sha256 } from '@noble/hashes/sha2';
import { bytesToHex } from '@noble/hashes/utils';

const enc = new TextEncoder();
function concat(a, b) {
  const o = new Uint8Array(a.length + b.length);
  o.set(a);
  o.set(b, a.length);
  return o;
}

export function pskFromToken(token) {
  return sha256(concat(enc.encode('terminal-relay/pair/psk'), token));
}
export function roomID(token) {
  return bytesToHex(sha256(concat(enc.encode('terminal-relay/pair/room'), token)).slice(0, 16));
}
export function encodeCode(signalURL, token) {
  const json = JSON.stringify({ s: signalURL, t: bytesToHex(token) });
  return btoa(json).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, ''); // base64url, no pad
}
export function decodeCode(code) {
  const b64 = code.replace(/-/g, '+').replace(/_/g, '/');
  const json = atob(b64);
  const p = JSON.parse(json);
  // Validate the token exactly like the Go reference (DecodeCode in code.go):
  // hex.DecodeString + len(token) != 16 -> "bad pairing code token".
  // A token of exactly 16 bytes is 32 hex chars; anything else fails closed
  // so a JS browser never silently derives a wrong/all-zeros psk + roomID.
  if (typeof p.t !== 'string' || !/^[0-9a-fA-F]{32}$/.test(p.t)) {
    throw new Error('bad pairing code token');
  }
  return { signalURL: p.s, token: hexToBytesLocal(p.t) };
}
function hexToBytesLocal(h) {
  const out = new Uint8Array(h.length / 2);
  for (let i = 0; i < out.length; i++) out[i] = parseInt(h.slice(i * 2, i * 2 + 2), 16);
  return out;
}
